pkg/util/liveconfig/dev.go

```diff
@@ -12,24 +12,45 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 )
 
-func (d *dev) HiveRestConfig(ctx context.Context, index int) (*rest.Config, error) {
-	// Indexes above 0 have _index appended to them
+func (d *dev) HiveRestConfig(ctx context.Context, shard int) (*rest.Config, error) {
+	// shards above 0 have _shard appended to it
 	envVar := hiveKubeconfigPathEnvVar
-	if index != 0 {
-		envVar = fmt.Sprintf("%s_%d", hiveKubeconfigPathEnvVar, index)
+	if shard != 0 {
+		envVar = fmt.Sprintf("%s_%d", hiveKubeconfigPathEnvVar, shard)
 	}
 
+	// Use an override kubeconfig path if one is provided
 	kubeConfigPath := os.Getenv(envVar)
-	if kubeConfigPath == "" {
-		return nil, fmt.Errorf("missing %s env variable", hiveKubeconfigPathEnvVar)
+	if kubeConfigPath != "" {
+		restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+		if err != nil {
+			return nil, err
+		}
+		return restConfig, nil
 	}
 
-	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	// Hive shards are planned but not deployed or implemented yet
+	d.hiveCredentialsMutex.RLock()
+	cached, exists := d.cachedCredentials[shard]
+	d.hiveCredentialsMutex.RUnlock()
+	if exists {
+		return rest.CopyConfig(cached), nil
+	}
+
+	// Lock the RWMutex as we're starting to fetch so that new readers will wait
+	// for the existing Azure API call to be done.
+	d.hiveCredentialsMutex.Lock()
+
+	kubeConfig, err := getAksKubeconfig(ctx, d.managedClustersClient, d.location, shard)
 	if err != nil {
+		d.hiveCredentialsMutex.Unlock()
 		return nil, err
 	}
 
-	return restConfig, nil
+	d.cachedCredentials[shard] = kubeConfig
+	d.hiveCredentialsMutex.Unlock()
+
+	return rest.CopyConfig(kubeConfig), nil
```

Comment:
In this time sensitive case it makes more sense to call the unlock deterministically at the exact moment it can be unlocked instead of relying on Golang to do it after the function exits.

Reply:
how much time are we saving by not using defer? it's a few nanoseconds at best, if this is the determining factor, we have much bigger problems. Unless there is data that shows that this is the problem in the monitor issue, I think we should use defer here since it's reducing code complexity.

Reply:
I missed those release notes, thanks for the background info. I've switched back to defer.